src/google-classroom-extractor/google_classroom_extractor/api/courses.py

```diff
@@ -119,18 +121,180 @@ def request_all_courses_as_df(
+
+    Parameters
+    ----------
+    courses_df: DataFrame
+        a Courses API DataFrame with the current fetched data which
+        will be mutated, adding Hash and CreateDate/LastModifiedDate
+    sync_db: sqlalchemy.engine.base.Engine
+        an Engine instance for creating database connections
+    """
+    courses_df.set_index('id', inplace=True)
+
+    # compute hash from API call data
+    courses_df["Hash"] = courses_df.apply(
+        lambda row: hashlib.sha256(
+            row.to_csv(header=False, index=False).encode("utf-8")
+        ).hexdigest(),
+        axis=1,
+    )
 
-    # append everything from API call
+    # add initial Create/Update times and SyncNeeded flag
+    now: datetime = datetime.now()
+    courses_df["CreateDate"] = now
+    courses_df["LastModifiedDate"] = now
+    courses_df["SyncNeeded"] = 1
+
+    # push to temporary sync table
     courses_df.to_sql(
-        "Courses", sync_db, if_exists="append", index=False, chunksize=500
+        "Sync_Courses", sync_db, if_exists="replace", index=True, chunksize=1000
     )
-    # remove duplicates - leave only the most recent
+
     with sync_db.connect() as con:
+        # ensure main table exists
         con.execute(
-            "DELETE from Courses "
-            "WHERE rowid not in (select max(rowid) "
-            "FROM Courses "
-            "GROUP BY id)"
+            "CREATE TABLE IF NOT EXISTS Courses ( "
+            "id BIGINT, "
+            "name TEXT, "
+            "section BIGINT, "
+            "descriptionHeading TEXT, "
+            "description TEXT, "
+            "room TEXT, "
+            "ownerId BIGINT, "
+            "creationTime TEXT, "
+            "updateTime TEXT, "
+            "enrollmentCode BIGINT, "
+            "courseState TEXT, "
+            "alternateLink TEXT, "
+            "teacherGroupEmail TEXT, "
+            "courseGroupEmail TEXT, "
+            "guardiansEnabled BOOLEAN, "
+            "calendarId BIGINT,  "
+            "Hash TEXT, "
+            "SyncNeeded BIGINT, "
+            "CreateDate DATETIME, "
+            "LastModifiedDate DATETIME, "
+            "PRIMARY KEY (id) "
+            ")"
         )
+        con.execute("CREATE INDEX IF NOT EXISTS SYNCNEEDED_INDEX ON Courses(SyncNeeded)")
 
-    return courses_df
+        # select unmatched records into temp table - differing by hash for same id
+        # single entry in result set if id only exists in one table (meaning add or missing),
+        #   so SyncNeeded flag will indicate which table it's from
+        # double entry in result set if id exists in both (meaning update needed),
+        #   so SyncNeeded will show which row is from which table
+        con.execute(
+            "CREATE TABLE Unmatched AS "
+            "SELECT * FROM ( "
+            "    SELECT * FROM Courses "
+            "    UNION ALL "
+            "    SELECT * FROM Sync_Courses "
+            ") "
+            "GROUP BY id, Hash "
+            "HAVING COUNT(*) = 1"
```

Comment:
Suggestion... easier to read and edit this way:
```suggestion
            """
            CREATE TABLE Unmatched AS
            SELECT * FROM (
                SELECT * FROM Courses
                UNION ALL
                SELECT * FROM Sync_Courses
            )
           GROUP BY id, Hash
           HAVING COUNT(*) = 1
           ''"
```

Reply:
👍